core/src/main/java/com/linecorp/armeria/server/annotation/JacksonResponseConverterFunction.java

```diff
@@ -66,23 +91,59 @@ public HttpResponse convertResponse(ServiceRequestContext ctx,
             if (mediaType.is(MediaType.JSON) || mediaType.subtype().endsWith("+json")) {
                 final Charset charset = mediaType.charset().orElse(StandardCharsets.UTF_8);
-
                 // Convert the object only if the charset supports UTF-8,
                 // because ObjectMapper always writes JSON document as UTF-8.
                 if (charset.contains(StandardCharsets.UTF_8)) {
-                    final HttpData content = HttpData.of(mapper.writeValueAsBytes(result));
-                    return HttpResponse.of(headers, content, trailingHeaders);
+                    if (result instanceof Publisher) {
+                        return aggregateFrom((Publisher<?>) result, headers, trailingHeaders,
+                                             this::toJsonHttpData);
+                    }
+                    if (result instanceof Stream) {
+                        return aggregateFrom((Stream<?>) result, headers, trailingHeaders,
+                                             this::toJsonHttpData, ctx.blockingTaskExecutor());
+                    }
+                    return HttpResponse.of(headers, toJsonHttpData(result), trailingHeaders);
                 }
             }
+
+            // @Produces("application/json-seq") or @ProducesJsonSequences is specified.
+            if (mediaType.is(MediaType.JSON_SEQ)) {
+                if (result instanceof Publisher) {
+                    return streamingFrom((Publisher<?>) result, headers, trailingHeaders,
+                                         this::toJsonSequencesHttpData);
+                }
+                if (result instanceof Stream) {
+                    return streamingFrom((Stream<?>) result, headers, trailingHeaders,
+                                         this::toJsonSequencesHttpData, ctx.blockingTaskExecutor());
+                }
+                return HttpResponse.of(headers, toJsonSequencesHttpData(result), trailingHeaders);
+            }
         } else if (result instanceof JsonNode) {
             // No media type is specified, but the result is a JsonNode type.
-            final HttpHeaders responseHeaders =
-                    headers.isImmutable() ? HttpHeaders.copyOf(headers).contentType(MediaType.JSON_UTF_8)
-                                          : headers.contentType(MediaType.JSON_UTF_8);
-            final HttpData content = HttpData.of(mapper.writeValueAsBytes(result));
-            return HttpResponse.of(responseHeaders, content, trailingHeaders);
+            return HttpResponse.of(toMutableHeaders(headers).contentType(MediaType.JSON_UTF_8),
+                                   toJsonHttpData(result), trailingHeaders);
         }
 
         return ResponseConverterFunction.fallthrough();
     }
+
+    private HttpData toJsonHttpData(@Nullable Object value) {
+        try {
+            return HttpData.of(mapper.writeValueAsBytes(value));
```

Comment:
I'm wondering if it's ok to produce the bytes which contain "null" string when the value is `null`.

Reply:
Yes, it's okay because `null` string is also a JSON value. :-)